tests/Functional/Client/FailureTestCase.php

```diff
@@ -80,4 +83,66 @@ public function testChildWorkflowFailurePropagation()
             $this->assertStringContainsString('SimpleActivity->fail', $e->getPrevious()->getMessage());
         }
     }
+
+    public function testSignalThatThrowsCustomError()
+    {
+        $client = $this->createClient();
+        $wf = $client->newWorkflowStub(SignalExceptionsWorkflow::class);
+
+        $run = $client->start($wf);
+
+        $wf->failWithName('test1');
+
+        try {
+            // The next
+            sleep(1);
+            $wf->exit();
+            $this->fail('Signal must fail');
+        } catch (AssertionFailedError $e) {
+            throw $e;
+        } catch (\Throwable $e) {
+            $this->assertInstanceOf(WorkflowNotFoundException::class, $e);
+            // \dump($e);
+        }
+
+        $this->expectException(WorkflowFailedException::class);
+        $result = $run->getResult();
+        $this->fail(sprintf("Workflow must fail. Got result %s", \print_r($result, true)));
+    }
+
+    public function testSignalThatThrowsInvalidArgumentException()
+    {
+        $client = $this->createClient();
+        $wf = $client->newWorkflowStub(SignalExceptionsWorkflow::class);
+
+        $run = $client->start($wf);
+
+        $wf->failInvalidArgument('test1');
+
+        $this->expectException(WorkflowFailedException::class);
+        $result = $run->getResult();
+        $this->fail(sprintf("Workflow must fail. Got result %s", \print_r($result, true)));
+    }
+
+    public function testSignalThatThrowsInternalException()
+    {
+        $client = $this->createClient();
+        $wf = $client->newWorkflowStub(SignalExceptionsWorkflow::class);
+
+        $run = $client->startWithSignal($wf, 'failActivity', ['foo']);
+
+        try {
+            sleep(5);
+            $wf->failActivity('foo');
+            $this->fail('Signal must fail');
+        } catch (AssertionFailedError $e) {
+            throw $e;
+        } catch (\Throwable $e) {
+            $this->assertInstanceOf(WorkflowNotFoundException::class, $e);
```

Comment:
Hmm... is there an `ActivityNotFoundException`? Seemingly this should throw that?

Reply:
Not really, this error is when we are trying to send a signal to an already dead workflow.

Reply:
Oooh, I see, I missed that it's already start-with-signal failing the activity earlier. Got it.